### ISSUE TITLE
Mintty discussion

### ISSUE THREAD

**Issue by mintty** [mintty/mintty#500]
As there is no active general discussion forum for mintty, I am reserving this issue for open discussion about further enhancements, opinions about open issues and their priorities and implementation options.

**Comment by DanielGGordon** [comment]
Why not use gitter.im?

**Comment by mintty** [comment]
Thanks for the hint. I have tentatively created a chat room.
Something like a *forum* (with structured topics) might be more useful, though.

**Comment by DanielGGordon** [comment]
I saw it and joined :)

I hear you on having 'topic' conversations. It looks like this is sort of possible with Gitter. Take a look at [FreeCodeCamp](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/Official-Free-Code-Camp-Chat-Rooms). They have multiple chat rooms for various topics.

It would be nice if there was some sort of a chat in Github projects. Right now the only real way to communicate (on Github itself) is to create an issue (like this), or to comment on a pull request/commit.

Sometimes I want to ask a 'stupid' question, or make a suggestion - but without making permanent clutter via an 'issue'.

Anyway I'm still fairly new to Github, so maybe it's just me.

**Comment by rdeforest** [comment]
Non-contributor here, chiming in a month and a half later...

My preference is to use GitHub's wiki feature for public non-chat team discussion. It has version control and markdown and doubles as documentation as participants refactor discussions. It doesn't automatically provide the thread structure a typical forum offers, but in my experience conversations in wikis are less prone to the kind of counter-productive banter one often sees in forum and email threads.

I see mintty already has seven wiki pages. If you're still looking for an option, add another page for "brainstorming" which lays down the barest of requirements of a topic and then supplies an example topic to use as a template.

**Comment by mintty** [comment]
Unfortunately I cannot add another page with different access rights. But I could set up a second wiki (e.g. github.com/mintty/forum/wiki). Do you have a few links how such wikis in forum-like usage might look and develop?

**Comment by mintty** [comment]
With a dedicated "forum" repository, I could also misuse the "issues" features to maintain a second, forum-like issues list, rather than a wiki. That would meet the desired use case.

**Comment by rdeforest** [comment]
For some reason I had this idea in my head that a github wiki was publicly editable by default. I should have checked first. Your idea of a separate project makes sense though.

It would be pretty impressive if I did have such links, but sadly no. The talk pages of any contested Wikipedia pages come to mind as examples. Discussion naturally adopts a tree-based format. When branches of the tree lose relevance it's kosher to prune them to an archive page. Of course, github's wiki doesn't permit "sub-pages", so users would have to maintain useful indexes.

"mintty/forum/wiki":
* Advantages
  * Setup cost is trivially low
  * Editing is almost identical to editing bugs
  * git-backed (can make massive changes atomically, rollback is easy)
  * users can clone the wiki repo and edit locally with their favorite editing tools
  * Re-factoring discussions is easier on a wiki than in structured discussions
* Disadvantages
  * Wikis of any kind require a minimum level of maturity, fastidiousness and discipline. In the absence of such the wiki can degenerate into a state which is actually worse than nothing.
    * No structure provided by default
    * Providing structure requires dev time which could be devoted to mintty instead

I hope my thoughts are useful. :)

**Comment by mintty** [comment]
My idea was actually to use the "issues" rather than the "wiki" feature of a dedicated "forum" repository, because it provides sorting by topics and threading implicitly.
But then there would be 2 issues domains, maybe confusing and not a good idea either.
I guess we'll rather wait until github provides a decent forum feature...

**Comment by cardiffman** [comment]
I'm just wondering how to sort of casually bring up that I get a -Werror error in config.c when I pull down and build the master branch using gcc 4.9.3 on cygwin. The error is about config.c line 676. I think the analysis that the variable is assigned and then not referenced is correct, when mis_config is not defined. 

I think it is valuable to be able to start by reporting something like this casually before filing a real issue.

**Comment by mintty** [comment]
Yes, that was a commit upload mistake to be fixed soon; note it's not a release, though.
Just remove the line locally for testing.

**Comment by mintty** [comment]
As an open design issue, for who is interested, I'd like to point out #524.

**Comment by mintty** [comment]
Request for comments: implementation details of drop mode #440.

**Comment by mintty** [comment]
See #568; I'm changing the bell to make sound by default, to comply with xterm.

**Comment by Equim-chan** [comment]
How should I donate mintty?

**Comment by mintty** [comment]
Thanks a lot for the offer. There is currently no need for donation but I'll think about a suitable way, maybe link to some charity.

**Comment by ariestav** [comment]
What are the license terms for including the build (.exe) of mintty in a commecial product?  I understand that the source is open, but is it possible to bundle the binary with a commercial product that is for sale to an end user?

**Comment by mintty** [comment]
License files are included in the repository. The main license for mintty is the GNU license. It has been causing me headaches already and as I'm not a lawyer, I'll not be trying to explain it to other people.
It would be nice to state, however, what product you have in mind.

**Comment by ariestav** [comment]
Yes, we develop a product called Templater.  It's an Adobe After Effects plugin.  You can learn more about the product here:  http://dataclay.com

Templater logs output to a text file and we want end-users to be able to `tail` their log files while the interact with the product's GUI.  The log files are UTF-8 encoded and with ANSI colors.  This works find out-of-the-box on macOS, however on Windows, there is no native `tail` command and color encoding does not seem to work as easily as it does on macOS.  Needless to say, development for this type of feature on Windows is difficult at best without something like mintty or Cygwin's terminal.  We are curious if we can ship a built binary of mintty in the product without penalty or fees. 

I will definitely read the license, but I'm not lawyer, either.  Any thoughts?

**Comment by mintty** [comment]
See https://www.gnu.org/licenses/gpl-faq.html, and note that you must also provide the sources of any GPL'ed code that you distribute.
Also note the file LICENSE.bundling in the mintty repository which I put up a while ago to satisfy a similar request.

**Comment by ariestav** [comment]
Thank you for this information.  I will investigate.  When it says "provide the source of any GPL'ed code that you distribute," does it suffice to point to this repository?  I would rather point to the official repo than to actually include the source code.  I understand if it's absolutely necessary to include the actual source code files, but is a reference to the source code repo acceptable?

**Comment by mintty** [comment]
> does it suffice to point to this repository?

Could be; that's how I do it myself.

**Comment by cardiffman** [comment]
I would point to a repo if it contains exactly what I build. If I made patches I would interpret the spirit of the GPL as, patches should be provided as well. Thus it may be more convenient to fork within GitHub and then point to the fork which has the patches.

**Comment by mintty** [comment]
Sounds OK, fork or separate patches as you prefer.
In either case, I would appreciate to get the purpose of patches explained, so if they are useful they can be integrated upstream.

**Comment by Biswa96** [comment]
Why does `mintty.exe -e conhost.exe` opens two console window and shows same text when I am writing some words?

**Comment by mintty** [comment]
With conhost from Windows (which you wouldn't normally run, must be a test case) or with a self-compiled thing? Cannot reproduce.

**Comment by Biswa96** [comment]
Open `cmd.exe` in the mintty folder. And run that command.

**Comment by mintty** [comment]
Apparently conhost is only visible from 64-bit Cygwin, not from 32-bit Cygwin.
When I run mintty -e conhost.exe (whether from mintty or from cmd) it only says "conhost: Exit 127."

**Comment by Biswa96** [comment]
I don't know if the following problem is for my case only. I open `mintty.exe --WSL` then run `ls` command in a big folder like `/usr/bin` or `/usr/sbin`. When the line reaches at end of window, the mintty windows crashes. Is it a bug?

**Comment by mintty** [comment]
Cannot reproduce. Is that with the release or the current repository version?
Crash means stall or close? Also with `-h always`? Maybe the shell terminates for some reason?

**Comment by Biswa96** [comment]
* When I open `mintty --WSL` and then run `ls /bin` the window close automatically without showing any error.
* When I open `mintty -h always --WSL` and then run `ls /bin` the window shows two errors `/bin/wslbridge: Exit 1.` and `wslbridge error: connection broke`. This happens when the command output is beyond the window size. May be this error not for mintty. Here is the screenshot:

![mintty_error_wsl](https://user-images.githubusercontent.com/31443074/36721738-c88540ae-1bd1-11e8-86c4-dfba222f6fa2.png)

**Comment by mintty** [comment]
Did you use a newer download of wslbridge than the latest release (assumedly to capture the mount configuration issue)? I had a similar problem before when I built the frontend and used the release backend which did not match: https://github.com/rprichard/wslbridge/issues/18#issuecomment-350643153

**Comment by Biswa96** [comment]
There are some interesting patches in msys2 mintty. See this link: https://github.com/Alexpux/MSYS2-packages/tree/master/mintty . You may add those patches 😋

**Comment by mintty** [comment]
Thanks for the hint;
* patch 0002-add-msys2-launcher.patch:

> Before, a pinned shortcut would have generic icon 

Pinning works in general; it uses the icon of the instance being pinned, I think. There is also the feature once provided for Git-Bash: Application Taskbar Shortcut Launch Command.

> and fail to launch a proper shell because of $MSYSTEM not being set.

This is an MSYS-specific issue, I think.

> The list of shells is harcoded.

If that relates to a "jump list": mintty has added that feature recently, and configurable.

* patch 0003-fix-current-dir-inheritance-for-alt-f2-on-msys2.patch:

There have been fixes about Alt+F2 in current dir in 2.8.0 and 2.8.1. If you clone an instance that invoked a login shell, the CHERE_INVOKING=1 might further help, but I'm not convinced that mintty should intrude so far into shell setup. Interworking with shell profiles needs to be discussed further, i.e. whether this is really a good idea.

**Comment by mintty** [comment]
@Biswa96, @Alexpux, @forgottenswitch, @JPeterMugaas:
The problem with CHERE_INVOKING is that it only applies to bash, not other shells. So this might cause rather suprising and unexpected inconsistencies. I don't think it's a good idea to intertwine terminal and shell configuration this way.
On the other hand, if someone sets CHERE_INVOKING before starting mintty, it will be inherited to further mintty sessions started with Alt+F2, so that would work already.

**Comment by Biswa96** [comment]
This is a end-user-type question: 
What are the difference between using mintty.exe and bash.exe in cygwin environment (except those shell, terminal, console etc. terms)?

**Comment by Biswa96** [comment]
I've seen that mintty gets wslbridge path from this line cmd = "/bin/wslbridge";` in winmain.c file. 

In Windows API one can get the full path of executable with `QueryFullProcessImageName()` and `GetCurrentProcess()`. Is it possible to add like that in POSIX?

For example, let the full path of mintty is `C:\foo\bar\mintty.exe`. So the result will be `wstring cmd = "C:\\foo\\bar\\wslbridge.exe` (not the .\\wslbridge).

**Comment by mintty** [comment]
/bin/wslbridge is a path in the Cygwin namespace; the Windows functions would not recognize it.
Also, what are you trying to achieve with this proposal? Find the wslbridge program? That works automatically within the Cygwin file system. The root is the parent of the folder where cygwin1.dll is loaded from. So /bin/wslbridge will be ../bin/wslbridge from there.

**Comment by Biswa96** [comment]
Why there is two mintty.exe after compiling it with cygwin? One in `/bin` another in `/bin/cygwin64`. 

In `winmain.c` file you've make a function `load_sys_library()` to find system32 folder and load DLL from that. But you may use `LoadLibraryEx` with `LOAD_LIBRARY_SEARCH_SYSTEM32` flag to find DLL from system32.

**Comment by mintty** [comment]
When I split the binary creation to be architecture-dependent (bin/cygwin32/ etc), I let it make a copy in bin/ to keep the cygport package deployment easy as before.

> LoadLibraryEx ...

I didn't write that function, has been untouched for years. "Never touch a running system."

**Comment by BrianInglis** [comment]
Not everyone running mintty uses Win10; some numbers of mintty users are still running XP and 7; some may use it under Wine: only LOAD_LIBRARY_... flag values 1,2,8 are supported on XP; many new values require patches for support on 7 and Vista.

**Comment by mintty** [comment]
@Biswa96, @Alexpux, @forgottenswitch, @JPeterMugaas:
I've changed my mind about setting CHERE_INVOKING even though it's shell-specific. But it makes sense to use an existing mechanism so that mintty can stay in the directory explicitly selected. So this is now applied if option `--dir` is used or if Alt+F2 clones the window session.

**Comment by mintty** [comment]
Mintty 2.8.5 opens Alt+F2 in same directory (if maintaining current directory information) even in login mode.

**Comment by mintty** [comment]
I'm closing this issue which has meanwhile become quite old, in favour of a new general chat issue #777.
Feel free to enter any remaining questions or proposals again there or in a new issue.
